api/test/opentelemetry/trace/tracer_test.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+
+describe OpenTelemetry::Trace::Tracer do
+  Propagation = OpenTelemetry::DistributedContext::Propagation
+  Trace = OpenTelemetry::Trace
+  Tracer = Trace::Tracer
+
+  # Tracer to verify expectation that `Span#finish` is called
+  class TestInSpanFinishTracer < Tracer
+    # Override `start_span` to return mock span
+    def start_span(*)
+      Minitest::Mock.new.expect(:finish, nil)
+    end
+  end
+
+  let(:invalid_span) { Trace::Span::INVALID }
+  let(:tracer) { Tracer.new }
+
+  describe '#current_span' do
+    let(:current_span) { tracer.start_span('current') }
+
+    it 'returns an invalid span by default' do
+      tracer.current_span.must_equal(invalid_span)
+    end
+
+    it 'returns the current span' do
+      wrapper_span = tracer.start_span('wrapper')
+
+      tracer.with_span(wrapper_span) do
+        tracer.current_span.must_equal(wrapper_span)
+      end
+    end
+  end
+
+  describe '#in_span' do
+    it 'yields the new span' do
+      tracer.in_span('wrapper') do |span|
+        span.wont_equal(invalid_span)
+        tracer.current_span.must_equal(span)
+      end
+    end
+
+    it 'returns the result of the block' do
+      result = tracer.in_span('wrapper') { 'my-result' }
+      result.must_equal('my-result')
+    end
+
+    it 'finishes the new span at the end of the block' do
+      finish_tracer = TestInSpanFinishTracer.new
+      mock_span = nil
+      finish_tracer.in_span('wrapper') { |span| mock_span = span }
+      mock_span.verify
+    end
+  end
+
+  describe '#with_span' do
+    it 'yields the passed in span' do
+      wrapper_span = tracer.start_span('wrapper')
+
+      tracer.with_span(wrapper_span) do |span|
+        span.must_equal(wrapper_span)
+      end
+    end
+  end
+
+  describe '#start_root_span' do
+    it 'requires a name' do
+      proc { tracer.start_root_span(nil) }.must_raise(ArgumentError)
+    end
+
+    it 'returns a valid span' do
+      span = tracer.start_root_span('root')
+      span.context.must_be :valid?
+    end
+  end
+
+  describe '#start_span' do
+    let(:context) { Trace::SpanContext.new }
+    let(:invalid_context) { Trace::SpanContext::INVALID }
+    let(:parent) { tracer.start_span('parent') }
+
+    it 'requires a name' do
+      proc { tracer.start_span(nil) }.must_raise(ArgumentError)
+    end
+
+    it 'returns a valid span' do
+      span = tracer.start_span('op', with_parent_context: context)
+      span.context.must_be :valid?
+    end
+
+    it 'returns a span with a new context by default' do
+      span = tracer.start_span('op')
+      span.context.wont_equal(tracer.current_span.context)
+    end
+
+    it 'returns a span with the same context as the parent' do
+      span = tracer.start_span('op', with_parent: parent)
+      span.context.must_equal(parent.context)
```

Comment:
This behaviour is wrong. Or maybe it isn't. It's definitely up for discussion, with the Named Tracers change. The SDK does not behave in this way, and I don't think the API's "no-op" implementation should either. Specifically, the new Span should share the parent span's trace ID, but should have a different span ID.

https://github.com/open-telemetry/opentelemetry-ruby/blob/3b52ae0beae634e6ab18db3b50c1bb4825b98c8f/api/lib/opentelemetry/trace/tracer.rb#L48-L50

vs.

https://github.com/open-telemetry/opentelemetry-ruby/blob/3b52ae0beae634e6ab18db3b50c1bb4825b98c8f/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L102

Reply:
Let's stick with the current behaviour for now. This matches the Java implementation of [`DefaultSpan`](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/DefaultSpan.java). There's some related debate in https://github.com/open-telemetry/opentelemetry-specification/issues/208